jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java

```diff
@@ -0,0 +1,281 @@
+    }
+
+    /**
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactor the exogenous factor
+     * @param maxIterations the maximum number of iterations to perform
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, double exogenousFactor, int maxIterations)
+    {
+        this(g, dampingFactor, exogenousFactor, maxIterations, TOLERANCE_DEFAULT);
+    }
+
+    /**
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactormap the exogenous factor map
+     * @param maxIterations the maximum number of iterations to perform
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, Map<V, Double> exogenousFactormap, int maxIterations)
+    {
+        this(g, dampingFactor, exogenousFactormap, maxIterations, TOLERANCE_DEFAULT);
+    }
+
+    /**
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactor the exogenous factor
+     * @param maxIterations the maximum number of iterations to perform
+     * @param tolerance the calculation will stop if the difference of AlphaCentrality values between
+     *        iterations change less than this value
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, double exogenousFactor, int maxIterations, double tolerance)
+    {
+        this.g = g;
+        this.scores = new HashMap<>();
+
+        if (maxIterations <= 0) {
+            throw new IllegalArgumentException("Maximum iterations must be positive");
+        }
+
+        if (dampingFactor < 0.0 || dampingFactor > 1.0) {
+            throw new IllegalArgumentException("Damping factor not valid");
+        }
+
+        if (tolerance <= 0.0) {
+            throw new IllegalArgumentException("Tolerance not valid, must be positive");
+        }
+
+        Map<V, Double> exofactor = new HashMap<>();
+        for (V v : g.vertexSet()) {
+            exofactor.put(v, exogenousFactor);
+            }
+
+        run(dampingFactor, exofactor, maxIterations, tolerance);
+    }
+
+    /**
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactormap the exogenous factor map
+     * @param maxIterations the maximum number of iterations to perform
+     * @param tolerance the calculation will stop if the difference of AlphaCentrality values between
+     *        iterations change less than this value
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, Map<V, Double> exogenousFactormap, int maxIterations, double tolerance)
```

Comment:
Similarly to what I described over at https://github.com/jgrapht/jgrapht/pull/501/files#r172380977, I suspect it's desirable for `exogenousFactormap` to be a `ToDoubleFunction<V>`, but I'd wait for the JGraphT team to comment on it first before changing anything.

Reply:
"exogenousFactormap" -> "exogenousFactor"?

Reply:
What I have done is I have used a function to look up to the map for fetching a value particularly related to a vertex. 
`ToDoubleFunction<Integer> i  = (x)-> f(x);` here your `f(x)` is the task you want to do when you receive x. So in our case, we have `V` instead of `Integer` as we are taking a vertex as input and so in case of scalar my `f(v)` is just the `exogenousfactor` and in case of map `f(v)` is `exogenousFactormap.get(v)` which corresponds to the exogenous factor value of that particular vertex.

Reply:
Ah, okay, I think I see what you've done.

My knowledge of alpha centrality is virtually nonexistent, so I don't have the full context here, but I still wonder though if it would be worth converting the `Map<V, Double>` parameter in this constructor to a `ToDoubleFunction<V>`.

The way I see it, it would give a user a lot more flexibility to pass in any way of converting vertices to exogenous factors that they can think of (so they aren't limited to just maps), and AFAICT it's much easier to convert a map to a function than the other way round.

(Compare:

```java
ToDoubleFunction<V> f = v -> ourExogenousFactorMap.get(v);
new AlphaCentrality(..., f, ...);
```

with

```java
Map<V, Double> m = ourVertices.stream().collect(toMap(v -> v, ourExogenousFactorFunction));
new AlphaCentrality(..., m, ...);
// or
Map<V, Double> m = ourVertices.stream().collect(toMap(v -> v, ourExogenousFactorFunction, (a, b) -> b, HashMap::new));
new AlphaCentrality(..., m, ...);
```

I'd argue the first option is more readable).

Not to mention that a `Map<V, Double>` would necessarily take more space and one is forced to box primitive `double`s to `Double` objects. :)

Reply:
@jbduncan The user does not want to convert vertices to exogenous factors but they want to assign each vertex a separate factor or they assign all the vertices the same factor by passing a scalar. Basically we are already giving the user a lot of flexibility by allowing them to pass a scalar or a map instead of converting the scalar to a map himself/herself.

Reply:
Okay. I admit that I'm not convinced by this argument, so I think it's best if we let the JGraphT team decide which route we should go. :)

Reply:
@jbduncan I think there is always a trade-off between certain things...in our case be it space for creating maps or allowing user to enter parameters other than maps. But I think if we refer to other standard libraries then they have implemented this particular algorithm for maps and scalars only. Moreover I think maps being fast should be preferred from user's point of view too.
I am really sorry if I am not able to convince you till now but would love to hear your points too.

Reply:
I tend to agree with @jbduncan.  Since JGraphT has been around since long before Java 8, a lot of our code does not take advantage of functional programming capabilities, but for new code, it's worth fresh thinking.

Reply:
@tibrewalpratik17 No worries!

Warning, and apologies in advance: wall of text incoming!

The only other argument I can think of in favour of `ToDoubleFunction<V>` over `Map<V, Double>` is that is better captures the semantics of what we're trying to do with this map.

To explain what I mean, in the `run` method of `AlphaCentrality`, the most that we're doing with the `ToDoubleFunction<V>` param there is we're querying it for an exogenous factor for a given vertex. In that sense, we're only really using the `Map#get` method of the passed-in map, rather than say `Map#put` as well. I don't imagine after this has all been reviewed that there'll be a need to depend on other methods of `Map` for what is effectively a function that accepts vertices and returns exogenous factors. Thus, the way I see it, if we tell the user through the API that they need to pass in a `java.util.Map`, then we're potentially saying that we need not just `Map#get` but other methods on `Map` as well, which AFAICT we don't.

So if a user has some other similar but sufficiently different data structure e.g. a [Vavr `Map`](https://www.javadoc.io/doc/io.vavr/vavr/0.9.2), then they've no choice but to convert their existing data structure to a `java.util.Map`, which may be costly in terms of either runtime performance or the mental effort needed to implement the code to do the conversion. Making the `Map` parameter a `ToDoubleFunction` would fix this problem.

Guava's graphs package approaches things similarly where for algorithms that only need to use one method of an input graph to do its job, e.g. `Graph#successors` for breadth first search, then the algorithm accepts e.g. a `SuccessorFunction` instead of a full-blown `Graph`, which allows e.g. users of JGraphT graphs to use Guava's graph algorithms. (I think this link to the [Guava wiki](https://github.com/google/guava/wiki/GraphsExplained#graph-types) explains things a bit better than I've done here.)

But having said all this, going back to the Vavr example, Vavr's `Map` does have a `toJavaMap` method which I believe is relatively efficient, so that particular example is a bit contrived.

**tl;dr:** `ToDoubleFunction` is more focused API-wise than `Map` since it has only one method, rather than N methods some of which could be used to mutate the passed-in `Map`, which we don't want to do.

Reply:
@jbduncan I went through the docs of [Successor Functions](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/graph/SuccessorsFunction.html) and [Predecessor Functions](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/graph/PredecessorsFunction.html) and really understood your point.

Reply:
@jsichi What will you suggest regarding it's implementation in this class?

Reply:
What @jbduncan suggested:  change the current constructor which takes a Map into one which takes a ToDoubleFunction.

Reply:
Done @jsichi.
Thanks @jbduncan. After implementing what you said I really got what exactly you were trying to point out. Basically converting a map or other data structures to a `ToDoubleFunction` is rather simple but vice-versa is not. Thus the user can simply convert his/her data structure into a `ToDoubleFunction` and pass as a parameter rather than to look for converting it into `Maps`. Sorry for all the trouble and I hope I got what you were saying this time correctly. Thanks a lot!!

Reply:
@tibrewalpratik16 You're welcome! 😀